src/workerd/api/worker-rpc.c++

```diff
@@ -148,64 +148,182 @@ private:
+  //
+  // Oh well. It's not a big deal. Just annoying that we have to forever support tearing RPC
+  // methods off their source object, even if we change implementations to something where that's
+  // less convenient.
+
+  kj::Vector<kj::StringPtr> path;
+  auto client = parent->getClientForOneCall(js, path);
+
   auto& ioContext = IoContext::current();
 
   auto builder = client.callRequest();
-  builder.setMethodName(name);
 
-  kj::Vector<jsg::JsValue> argv(args.Length());
-  for (int n = 0; n < args.Length(); n++) {
-    argv.add(jsg::JsValue(args[n]));
+  if (path.empty()) {
+    builder.setMethodName(name);
+  } else {
+    auto pathBuilder = builder.initMethodPath(path.size() + 1);
+    for (auto i: kj::indices(path)) {
+      pathBuilder.set(i, path[i]);
+    }
+    pathBuilder.set(path.size(), name);
   }
 
-  // If we have arguments, serialize them.
-  // Note that we may fail to serialize some element, in which case this will throw back to JS.
-  if (argv.size() > 0) {
-    serializeJsValue(js, js.arr(argv.asPtr()), [&](capnp::MessageSize hint) {
-      // TODO(perf): Actually use the size hint.
-      return builder.initArgs();
-    });
-  }
+  fillOpFunc(builder.getOperation());
 
   auto callResult = builder.send();
 
-  return ioContext.awaitIo(js, kj::mv(callResult),
-      [](jsg::Lock& js, auto result) -> jsg::Value {
-    return jsg::Value(js.v8Isolate, deserializeJsValue(js, result.getResult()));
+  return {
+    .promise = jsg::JsPromise(js.wrapSimplePromise(ioContext.awaitIo(js, kj::mv(callResult),
+        [](jsg::Lock& js, auto result) -> jsg::Value {
+      return jsg::Value(js.v8Isolate, deserializeJsValue(js, result.getResult()));
+    }))),
+    .pipeline = kj::mv(callResult),
+  };
+}
+
+jsg::Ref<JsRpcPromise> JsRpcProperty::call(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  jsg::Lock& js = jsg::Lock::from(args.GetIsolate());
+
+  auto [promise, pipeline] = callImpl(js, [&](rpc::JsRpcTarget::CallParams::Operation::Builder op) {
+    kj::Vector<jsg::JsValue> argv(args.Length());
+    for (int n = 0; n < args.Length(); n++) {
+      argv.add(jsg::JsValue(args[n]));
+    }
+
+    // If we have arguments, serialize them.
+    // Note that we may fail to serialize some element, in which case this will throw back to JS.
+    if (argv.size() > 0) {
+      serializeJsValue(js, js.arr(argv.asPtr()), [&](capnp::MessageSize hint) {
+        // TODO(perf): Actually use the size hint.
+        return op.initCallWithArgs();
+      });
+    }
   });
+
+  return jsg::alloc<JsRpcPromise>(
+      jsg::JsRef<jsg::JsPromise>(js, promise),
+      IoContext::current().addObject(kj::heap(kj::mv(pipeline))));
+}
+
+namespace {
+
+jsg::JsValue thenImpl(jsg::Lock& js, v8::Local<v8::Promise> promise,
+      v8::Local<v8::Function> handler, jsg::Optional<v8::Local<v8::Function>> errorHandler) {
+  KJ_IF_SOME(e, errorHandler) {
+    return jsg::JsPromise(jsg::check(promise->Then(js.v8Context(), handler, e)));
```

Comment:
The use of `jsg::check(...)` here will call these to throw synchronously... which I think is what we probably want in the cases that would normally cause `->Then(...)` to fail. But it's worth double checking, and likely worth a comment in here since normally we would expect `then()`, `catch()` and `finally()` to catch exceptions and turn them into rejections.

Reply:
If this throws here, it's because calling `.then()` on the underlying V8 promise threw synchronously. If the V8 promise threw synchronously then our wrapper presumably should too.

But I should make sure KJ exceptions thrown synchronously from capnp are caught and turned async...

Reply:
> If this throws here, it's because calling .then() on the underlyuing V8 promise threw...

Yep. As I said, likely worth a brief comment here.